File: razer-sdk/src/main/java/com/bonepl/razersdk/color/MultiTransitionColor.java
package com.bonepl.razersdk.color;

import java.util.LinkedList;
import java.util.List;

public class MultiTransitionColor implements Color {
    private final List<StaticColor> colorsList;
    private int i = 0;

    private MultiTransitionColor(List<StaticColor> colorsList) {
        this.colorsList = colorsList;
    }

    @Override
    public StaticColor getColor() {
        if (i == colorsList.size()) {
            i = 0;
        }
        return colorsList.get(i++);
    }

    public int getTotalTransitions() {
        return colorsList.size();
    }

    public static class Builder {
        private final LinkedList<StaticColor> builderList = new LinkedList<>();
        private int loopedSteps;

        public Builder(StaticColor startingColor) {
            builderList.add(startingColor);
        }

        public Builder addTransition(StaticColor toColor, int steps) {
            if (steps < 2) {
                throw new IllegalArgumentException("Minimum amount of steps in transition color is 2, requested: " + steps);
            }
            TransitionColor transitionColor = new TransitionColor(builderList.getLast(), toColor, steps);
            //skip first color as it is already on list
            transitionColor.getColor();
            while (!transitionColor.transitionFinished()) {
                builderList.add(transitionColor.getColor());
            }
            return this;
        }

        public Builder looped(int loopedSteps) {
            this.loopedSteps = loopedSteps;
            return this;
        }

        public MultiTransitionColor build() {
            if (loopedSteps != 0) {
                addTransition(builderList.getFirst(), loopedSteps);
                builderList.removeLast();
            }
            return new MultiTransitionColor(builderList);
        }
    }
}

File: src/razer-sdk/color/MultiTransitionColor.ts
import {Color} from "./Color.js";
import {StaticColor} from "./StaticColor.js";
import {TransitionColor} from "./TransitionColor.js";

export class MultiTransitionColor implements Color {
    private readonly colorsList: StaticColor[];
    private i: number = 0;

    private constructor(colorsList: StaticColor[]) {
        this.colorsList = colorsList;
    }

    getColor(): StaticColor {
        if (this.i === this.colorsList.length) {
            this.i = 0;
        }
        return this.colorsList[this.i++];
    }

    getTotalTransitions(): number {
        return this.colorsList.length;
    }

    static Builder = class {
        readonly builderList: StaticColor[] = [];
        loopedSteps: number = 0;

        constructor(startingColor: StaticColor) {
            this.builderList.push(startingColor);
        }

        addTransition(toColor: StaticColor, steps: number): this {
            if (steps < 2) {
                throw new Error(
                    "Minimum amount of steps in transition color is 2, requested: " + steps
                );
            }
            const transitionColor = new TransitionColor(
                this.builderList[this.builderList.length - 1],
                toColor,
                steps
            );
            // skip first color as it is already on list
            transitionColor.getColor();
            while (!transitionColor.transitionFinished()) {
                this.builderList.push(transitionColor.getColor());
            }
            return this;
        }

        looped(loopedSteps: number): this {
            this.loopedSteps = loopedSteps;
            return this;
        }

        build(): MultiTransitionColor {
            if (this.loopedSteps !== 0) {
                this.addTransition(this.builderList[0], this.loopedSteps);
                this.builderList.pop();
            }
            return new MultiTransitionColor(this.builderList);
        }
    };
}
